packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioLibraryPublicationHandler.java

```diff
@@ -0,0 +1,264 @@
+                        Resource libraryResource = rSet.createResource(new JSONResourceFactory().createResourceURI(UUID.nameUUIDFromBytes(resourceId.getBytes()).toString()));
+                        libraryResource.getContents().add(source);
+                        // Move dependencies that won't produce a library into the resource.
+                        for (EObject dependency : dependencyEntry.getValue()) {
+                            if (!this.isLibraryCandidate.test(dependency)) {
+                                libraryResource.getContents().add(dependency);
+                            }
+                        }
+                    }
+
+                }
+            }
+
+            // Cache the created libraries to resolve inter-library dependencies.
+            Map<EObject, Library> libraries = new HashMap<>();
+
+            for (EObject source : dependencies.keySet()) {
+                if (this.isLibraryCandidate.test(source)) {
+                    Optional<String> optionalContent = this.toJson(source.eResource());
+                    Optional<String> optionalName = this.getLibraryName(source);
+                    if (optionalContent.isPresent() && optionalName.isPresent()) {
+                        long librariesWithSameNameCount = libraries.values().stream().filter(library -> Objects.equals(library.getName(), optionalName.get())).count();
+                        String name = optionalName.get();
+                        if (librariesWithSameNameCount > 0) {
+                            // Handle duplicated names by suffixing an index
+                            name = name + librariesWithSameNameCount;
+                        }
+                        Library library = Library.newLibrary()
+                                .project(AggregateReference.to(input.projectId()))
+                                .name(name)
+                                .content(optionalContent.get())
+                                .version(input.version())
+                                .description(input.description())
+                                .build(input);
+                        libraries.put(source, library);
+                    }
+                }
+            }
+
+            // Set the dependencies between libraries.
+            for (Map.Entry<EObject, Set<EObject>> dependencyEntry : dependencies.entrySet()) {
+                Library sourceLibrary = libraries.get(dependencyEntry.getKey());
+                for (EObject dependency : dependencyEntry.getValue()) {
+                    Library dependencyLibrary = libraries.get(dependency);
+                    if (dependencyLibrary != null) {
+                        sourceLibrary.addDependency(input, dependencyLibrary.getId());
+                    }
+                }
+            }
+            persistedLibraries = this.libraryRepository.saveAll(libraries.values());
```

Comment:
I'd like to return a `Failure` if the creation of the libraries produces a duplicated key (same version, name, and projectId). How should I do that?
- Let the database layer throw an exception and catch it to create my `Failure`?
- Check if a library with the given name, version, and projectId exists and create the `Failure` before attempting to persist?